WordPress-iOS-Shared/Core/OperationPools.swift

```diff
@@ -0,0 +1,286 @@
+        }
+    }
+}
+
+/// An OperationPool that transforms a NSURLRequest into a Value.
+///
+/// It uses `NSURLSession.downloadTaskWithRequest` to download the file for the
+/// given request, then uses a `Parser` to convert the downloaded file into the
+/// wanted `Value`.
+///
+public class NetworkDownloadPool<Value>: OperationPool<NSURLRequest, Value, NSError> {
+    public typealias Parser = NSURL -> Result<Value, NSError>
+
+    /// Constructs a new NetworkDownloadPool with the given parser.
+    ///
+    /// Optionally, you can specify the NSURLSession to use. By default, it uses
+    /// the shared session.
+    ///
+    /// - parameter parser: A function to be called to convert the downloaded
+    /// file to a `Value`. The `parser` function takes a file `NSURL` that
+    /// points to the location of the downloaded file, and returns a `Result`
+    /// with the processed `Value`, or a `NSError`.
+    ///
+    /// - parameter session: The NSURLSession to use for the network request. By
+    /// default it uses `NSURLSession.sharedSession()`.
+    ///
+    public init(parser: Parser, session: NSURLSession = NSURLSession.sharedSession()) {
+        super.init { request, completion in
+            let task = session.downloadTaskWithRequest(request) {
+                location, response, error in
+
+                let imageLocation = Result(value: location, error: error, fallbackError: OperationPoolError.noDataAndNoError.nserror)
+                let image = imageLocation.flatMap(parser)
+                completion(image)
+            }
+            task.resume()
+        }
+    }
+}
+
+/// An OperationPool to download images.
+///
+/// Builds on NetworkDownloadPool and transforms a `NSURLRequest` into a
+/// `UIImage`.
+///
+/// If the downloaded data can't be turned into an image, it results in a
+/// `OperationPoolError.invalidImageData` error.
+///
+public class ImageDownloadPool: NetworkDownloadPool<UIImage> {
+
+    /// Constructs a new ImageDownloadPool.
+    ///
+    /// Optionally, you can specify the NSURLSession to use. By default, it uses
+    /// the shared session.
+    ///
+    /// - parameter session: The NSURLSession to use for the network request. By
+    /// default it uses `NSURLSession.sharedSession()`.
+    ///
+    public init(session: NSURLSession = NSURLSession.sharedSession()) {
+        let parser: Parser = { url in
+            do {
+                let data = try NSData(contentsOfURL: url, options: .DataReadingUncached)
+                if let image = UIImage(data: data) {
+                    return .success(image)
+                } else {
+                    return .failure(OperationPoolError.invalidImageData.nserror)
+                }
+            } catch {
+                return .failure(error as NSError)
+            }
+        }
+        super.init(parser: parser, session: session)
+    }
+}
+
+/// Errors related to OperationPool.
+/// 
+public enum OperationPoolError: Int, ErrorType {
```

Comment:
I removed the `ErrorType` last night but then reverted it. In Swift, `ErrorType` is an empty protocol, which just means that a type represents some kind of error.

Swift automatically bridges ErrorType to NSError, but it's not publicly exposed so we can't override how it does it.

![screen shot 2016-04-20 at 08 39 45](https://cloud.githubusercontent.com/assets/8739/14665428/774ffb08-06d3-11e6-96f4-ca5fddd28128.png)

I think a good idea would be to expose `domain`/`code`. If we wanted to check for a specific error, with the current version:
```swift
pool.request(request) { result in
    if let error = result.error
        where error == OperationPoolError.noDataAndNoError.nserror {
        // handle specific error
    }
}
```

Exposing domain/code:

```swift
pool.request(request) { result in
    if let error = result.error
        where error.domain == OperationPoolError.domain
            && error.code == OperationPoolError.noDataAndNoError.code {
        // handle specific error
    }
}
```

I think I prefer the first version though, as it's more direct in its comparison.

An even more _classic_ approach would be to define the error enum like this:

```swift
public enum OperationPoolError {
    public static let domain = "org.wordpress.shared.operationpool"
    public enum Code: Int {
        case noDataAndNoError
        case invalidImageData
    }

    private static var noDataAndNoError: NSError {
        return NSError(domain: domain, code: Code.noDataAndNoError.rawValue, userInfo: [NSLocalizedDescriptionKey: "Network response had no data and no error"])
    }
    private static var invalidImageData: NSError {
        return NSError(domain: domain, code: Code.invalidImageData.rawValue, userInfo: [NSLocalizedDescriptionKey: "Couldn't generate a valid image with the returned data"])
    }
}
```

And this is how you'd check:

```swift
pool.request(request) { result in
    if let error = result.error
        where error.domain == OperationPoolError.domain
            && error.code == OperationPoolError.Code.noDataAndNoError.rawValue {
        // handle specific error
    }
}
```

But checking for an error gets even more verbose.